### ISSUE TITLE
Training Ineffective (mAP ≈ 0) When Number of Training Images ≤ Batch Size

### ISSUE THREAD

**Issue by devansh-lodha** [ultralytics/ultralytics#20663]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

When training a YOLO model (e.g.`yolo11s.pt`) on a custom dataset, the model achieves a mean Average Precision (mAP50-95) score that is effectively zero if the number of unique training images (`N_imgs`) is less than or equal to the specified `batch_size`.
Once `N_imgs` becomes strictly greater than `batch_size`, the mAP50-95 score shows a significant jump, indicating that learning starts to occur as expected. This behavior has been observed consistently across different `batch_size` values (tested systematically with 32 and 128), with the "jump point" for mAP scaling accordingly.

This suggests a potential issue in the data loading mechanism or the training loop that prevents effective training when the total number of training samples cannot form more than one full batch, or is otherwise mishandled when `N_imgs <= batch_size`.

**Expected Behavior:**
The model should demonstrate some level of learning (mAP50-95 > ~0.0 and generally increasing with `N_imgs`) even when `N_imgs <= batch_size`, as long as `N_imgs` is at least 1. While performance is expected to be suboptimal with very few images, it should not be consistently near-zero until `N_imgs` strictly exceeds `batch_size`. The training process should effectively utilize all provided training images.

**Actual Behavior (Summary from more extensive tests):**
The mAP50-95 remains near zero until `N_imgs > batch_size`.

**Test 1: `batch_size = 32`, `epochs = 10`, `model = yolo11s.pt` (using Ultralytics `8.3.134`)**
| Num Train Images | mAP50-95 |
|------------------|----------|
| 1                | 0.0006   |
| 2                | 0.0007   |
| 4                | 0.0010   |
| 8                | 0.0007   |
| 15               | 0.0009   |
| 16               | 0.0008   |
| 30               | 0.0007   |
| 31               | 0.0007   |
| 32 (== BS)       | 0.0007   |
| **33 (> BS)**    | **0.0130**   |
| 34               | 0.0164   |
| 40               | 0.0526   |
| 64               | 0.0791   |
| 96               | 0.2272   |
| 128              | 0.3715   |

![Image](https://github.com/user-attachments/assets/6a6a95f2-8ae2-4736-a91d-19f9f1c4ea3f)

**Test 2: `batch_size = 128`, `epochs = 10`, `model = yolo11s.pt` (using Ultralytics `8.3.134`)**
| Num Train Images | mAP50-95 |
|------------------|----------|
| 1                | 0.0012   |
| 16               | 0.0008   |
| 32               | 0.0007   |
| 64               | 0.0006   |
| 96               | 0.0006   |
| 120              | 0.0006   |
| 127              | 0.0006   |
| 128 (== BS)      | 0.0006   |
| **129 (> BS)**   | **0.0062**   |
| 130              | 0.0106   |
| 140              | 0.0560   |
| 160              | 0.0763   |
| 192              | 0.0890   |
| 256              | 0.0704   |

![Image](https://github.com/user-attachments/assets/ae3d4b6f-efda-4206-822b-a1489d828e60)

As can be seen, there is a distinct "jump" in mAP50-95 only when the number of training images exceeds the batch size. Even when `N_imgs == batch_size`, the performance remains at near-zero levels.

This behavior could lead users with very small datasets (or those performing initial sanity checks with a tiny subset of images) to incorrectly conclude that their setup or data is entirely non-functional.

### Environment

```
Ultralytics 8.3.134 🚀 Python-3.13.2 torch-2.7.0+cu126 CUDA:0 (NVIDIA A100-SXM4-80GB, 81156MiB)
Setup complete ✅ (64 CPUs, 503.7 GB RAM, 199.7/430.7 GB disk)

OS                  Linux-5.15.0-139-generic-x86_64-with-glibc2.31
Environment         Linux
Python              3.13.2
Install             pip
RAM                 503.73 GB
Disk                199.7/430.7 GB
CPU                 AMD EPYC 7452 32-Core Processor
CPU count           64
GPU                 NVIDIA A100-SXM4-80GB, 81156MiB
GPU count           4
CUDA                12.6

numpy               ✅ 2.2.5>=1.23.0
matplotlib          ✅ 3.10.3>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 11.2.1>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.15.3>=1.4.1
torch               ✅ 2.7.0>=1.8.0
torch               ✅ 2.7.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.22.0>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 7.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.14>=2.0.0
```

### Minimal Reproducible Example

**Methodology Overview:**
1.  **Dataset:** A public object detection dataset from Roboflow ("liangdianzhong/-qvdww/3", 911 training images, 23 classes, YOLO format) was used as the base. Any standard object detection dataset could be used. A consistent validation set was used for all training runs.
2.  **Model:** `yolo11s.pt` (as downloaded by Ultralytics `8.3.134`) was used for fine-tuning. 
3.  **Experiment Loop:**
    *   A fixed `batch_size` was chosen (e.g., 32, then 128).
    *   The number of training images (`N_imgs`) was varied across a predefined list (e.g., [1, ..., 31, 32, 33, ..., 128] for `batch_size=32`).
    *   For each `N_imgs`:
        *   A random subset of `N_imgs` images and their corresponding label files were copied from the original training set into a temporary directory.
        *   A temporary `data_subset.yaml` was created, pointing to this subset for `train` and to the original, consistent validation set for `val`.
        *   The model was trained from the pre-trained weights using the `yolo` CLI:
            ```bash
            yolo task=detect mode=train model=yolo11s.pt data=<path_to_subset_yaml> epochs=10 imgsz=640 batch=<current_batch_size> patience=0 name=train_N<N_imgs>_E10_B<batch_size> plots=True exist_ok=True
            ```
            (`patience=0` was used to ensure all 10 epochs completed for fair comparison).
        *   The final `mAP50-95 (Box)` was recorded from the `results.csv` file generated by the training run.
4.  **Analysis:** The recorded `mAP50-95` values were plotted against `N_imgs` for each `batch_size` tested.

**Link to GitHub Gist with Full Experimental Script and Setup Details:**

```
https://gist.github.com/devansh-lodha/e026368c4bd45b8bda7acaaa6cfcd8b7
```

**To Reproduce using the Gist:**
1.  Clone the Gist or copy the Python script.
2.  Ensure you have the necessary environment (Python, Ultralytics, Roboflow library - see "Environment" section of this bug report and `requirements.txt` included in Gist).
3.  Set up your Roboflow API key if you intend to download the same dataset (or adapt the script's data loading part for your own local dataset).
4.  Run the Python script. It will:
    *   Download the dataset (if configured for Roboflow).
    *   Create the described training subsets.
    *   Execute the YOLO training commands for different `N_imgs` values for one or more batch sizes.
    *   Collect and plot the results, which should demonstrate the mAP jump described.

The script is configured to test with `batch_size = 32` and then `batch_size = 128` (these can be adjusted in the script). The key observation will be the mAP50-95 scores being near zero until `N_imgs` exceeds the `batch_size`.

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @devansh-lodha, thank you for the detailed report and for your interest in Ultralytics 🚀! This is an automated response to help streamline the troubleshooting process. An Ultralytics engineer will also assist you here soon.

We recommend checking the [Docs](https://docs.ultralytics.com/) for additional [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples, as well as answers to many frequently asked questions.

Since this is a 🐛 Bug Report, thank you for including a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/)—that’s extremely helpful for our debugging process!

If you have any further details, logs, or clarifications to add, please include them here to assist our team. Also, ensure you’re following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) to optimize your experiments.

Join the Ultralytics community where it suits you best:
- Real-time chat: [Discord](https://discord.com/invite/ultralytics) 🎧
- In-depth discussions: [Discourse](https://community.ultralytics.com/)
- Community threads: [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

To help us rule out issues that may already be resolved, please upgrade to the latest `ultralytics` package along with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you again for your detailed report and willingness to contribute! One of our engineers will review and respond as soon as possible.

**Comment by Y-T-G** [comment]
Did you try with `nbs=1`?

**Comment by devansh-lodha** [comment]
Hi @Y-T-G ,

Thank you so much for the suggestion to try `nbs=1`! This made sense to me because with the default `nbs=64` (Nominal Batch Size), the trainer accumulates gradients and only performs an optimizer step after processing 64 images. If the total number of training images (`N_imgs`) in a subset is less than 64, an optimizer step might not be performed within an epoch, potentially leading to no weight updates and thus no learning.

I have re-run the experiments for both `batch_size=32` and `batch_size=128` using the following parameters in the `yolo` training command for each run:
*   `nbs=1`
*   `cache=False`
*   `seed=0`
*   `epochs=10`
*   `patience=0`
*   `model=yolo11s.pt`

**Test Results: `batch_size = 32`, `nbs=1`**

| Num Train Images | mAP50-95   |
|------------------|------------|
| 1                | 0.0007     |
| 2                | 0.0011     |
| 4                | 0.0010     |
| 8                | 0.0006     |
| 15               | 0.0008     |
| 16               | 0.0008     |
| 30               | 0.0008     |
| 31               | 0.0008     |
| 32 (== BS)       | 0.0008     |
| **33 (> BS)**    | **0.0233** |
| 34               | 0.0258     |
| 40               | 0.0674     |
| 64               | 0.0796     |
| 96               | 0.2454     |
| 128              | 0.3657     |

![Image](https://github.com/user-attachments/assets/b30c431c-04bc-4ffc-b237-5c2937495a0a)

**Test Results: `batch_size = 128`, `nbs=1`**

| Num Train Images | mAP50-95   |
|------------------|------------|
| 1                | 0.0007     |
| 16               | 0.0009     |
| 32               | 0.0007     |
| 64               | 0.0006     |
| 96               | 0.0005     |
| 120              | 0.0006     |
| 127              | 0.0005     |
| 128 (== BS)      | 0.0006     |
| **129 (> BS)**   | **0.0060** |
| 130              | 0.0163     |
| 140              | 0.0651     |
| 160              | 0.0706     |
| 192              | 0.0874     |
| 256              | 0.0914     |

![Image](https://github.com/user-attachments/assets/9868647b-e136-43ac-800c-3261d6abe895)

**Key Observations with `nbs=1`:**
1.  mAP50-95 remains near-zero when `N_imgs <= batch_size`.
2.  A distinct and significant mAP jump occurs only when `N_imgs > batch_size`.

This persistence of the issue, even when forcing optimizer updates after every processed batch (due to `nbs=1`), suggests the problem is not solely related to gradient accumulation tied to `nbs > 1`.
